build.rs

```diff
@@ -0,0 +1,214 @@
+        for account_type in &self.registry {
+            if let Some(clash) = used_prefixes.insert(account_type.prefix, (*account_type).clone())
+            {
+                return Err(format!(
+                    "prefixes must be unique but this account's prefix:\n{:#?}\nclashed with\n{:#?}",
+                    account_type,
+                    clash
+                ));
+            }
+            if let Some(clash) = used_networks.insert(account_type.name(), account_type.clone()) {
+                return Err(format!(
+                    "networks must be unique but this account's network:\n{:#?}\nclashed with\n:{:#?}",
+                    account_type,
+                    clash
+                ));
+            }
+            if account_type.network.is_empty() {
+                return Err("network is mandatory.".into());
+            }
+
+            if let Err(err) = is_valid_rust_identifier(&account_type.name()) {
+                return Err(format!(
+                    "network not valid: {} for {:#?}",
+                    err, account_type
+                ));
+            }
+            if account_type.decimals.len() != account_type.symbols.len() {
+                return Err(format!(
+                    "decimals must be specified for each symbol: {:?}",
+                    account_type
+                ));
+            }
+            if let Some(sig_type) = &account_type.standard_account {
+                match sig_type.as_str() {
+                    "Sr25519"| "Ed25519" | "secp256k1" | "*25519" => {},
+                    _ => {
+                        return Err(format!("Unknown sig type in standardAccount: expected one of Sr25519, Ed25519, secp256k1, *25519: {:?}", account_type))
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+struct AccountType {
+    prefix: u16,
+    network: String,
+    display_name: String,
+    /// If standard account is None then the network is reserved.
+    standard_account: Option<String>,
+    symbols: Vec<String>,
+    decimals: Vec<u8>,
+    website: Option<String>,
+}
+
+impl AccountType {
+    fn name(&self) -> String {
+        format!(
+            "{}Account",
+            inflector::cases::pascalcase::to_pascal_case(&self.network)
+        )
+    }
+
+    fn is_reserved(&self) -> bool {
+        self.standard_account.is_none()
+    }
+}
+
+fn create_ss58_registry(json: &str) -> Result<proc_macro2::TokenStream, String> {
+    let registry: Registry = serde_json::from_str(json).expect("valid json file");
+    registry.is_valid()?;
+
+    let mut registry = registry.registry;
+    // The names are assumed to be sorted.
+    registry.sort_by_key(|a| a.name());
+
+    // Variables to insert into quote template:
+    let identifier: Vec<_> = registry
+        .iter()
+        .map(|r| format_ident!("{}", r.name()))
+        .collect();
```

Comment:
Do we need the collect here?

Reply:
We need a vec as it's read multiple times.